### ISSUE TITLE
Filesystem Input to get filesystem metrics

### ISSUE THREAD

**Issue by hdiass** [fluent/fluent-bit#681]
**Is your feature request related to a problem? Please describe.**
No, this is feature request.

**Describe the solution you'd like**
Would be very useful that the disk usage input could get more details on disks attached to the computational unit(vm or not).

Metrics to export:
The disk space available to an unprivileged user in bytes.

The disk name. For example: /dev/disk1

The disk type. For example: ext4

The mounting point. For example: /

The total number of file nodes in the file system.

The disk space available in bytes.

The number of free file nodes in the file system.

The total disk space in bytes.

The used disk space in bytes.

The percentage of used disk space.

Some fields are not that relevant for my needs, but could be for other purposes.

Thanks in advance,
HD

**Comment by fopina** [comment]
Even though this is 1y old I'll bump this as I was also looking for it.
Sadly plugin system is only for outputs, so one would have to do get this merged into master to be able to use the official release... I guess this is the only gap to fluentd.

In the meantime, as there is `exec` input now, I've worked around it using `stat` and a custom parser.

```ini
[PARSER]
    Name    disk-stats
    Format  regex
    Regex   \s+(?<id>[a-z0-9]+)\s+(?<nl>\d+)\s+(?<type>[a-z0-9]+)\s+(?<bs>\d+)\s+(?<fbs>\d+)\s+(?<total>\d+)\s+(?<free>\d+)\s+(?<avail>\d+)\s+(?<itotal>\d+)\s+(?<ifree>\d+)
```
```
/opt/td-agent-bit/bin/td-agent-bit -R /etc/td-agent-bit/parsers.conf -i exec -p command='stat -tf /' -pParser=disk_stats -o stdout
...
[0] exec.0: [1562941623.012136101, {"id"=>"1337324b7279c477", "nl"=>"255", "type"=>"ef53", "bs"=>"4096", "fbs"=>"4096", "total"=>"3797887", "free"=>"2971402", "avail"=>"2807235", "itotal"=>"950096", "ifree"=>"817145"}]
```
Raw `stat` data gets confusing in dashboards due to different blocksizes, but in my case it's not relevant and prefer not adding the overhead of piping to `bc` or doing a new stat-like binary just for that

**Comment by nigels-com** [comment]
Here is the busybox implementation of `df` for inspiration.
https://git.busybox.net/busybox/tree/coreutils/df.c

The good news is that it's not all that complicated to get a msgpack/JSON representation for POSIX/LINUX, at least.

* Read `/proc/mounts` to get the list of mounts
* Use [statvfs](http://man7.org/linux/man-pages/man3/statvfs.3.html) for each mount
* Skip the ones with no blocks
* Perhaps skip the read-only mounts
* Perhaps filter out various system mounts

```
$ cat /proc/mounts | grep -v '^overlay' | grep -v '^/dev/loop' | grep -v '^cgroup ' | grep -v '^shm ' | grep -v '^nsfs '
```

**Comment by nigels-com** [comment]
Also, busybox [stat](https://git.busybox.net/busybox/tree/coreutils/stat.c)

```
$ stat -f /
  File: "/"
    ID: de80f3606b5ddd12 Namelen: 255     Type: ext2/ext3
Block size: 4096       Fundamental block size: 4096
Blocks: Total: 23898985   Free: 4821451    Available: 3596671
Inodes: Total: 6111232    Free: 3372365
```

**Comment by markstos** [comment]
You can also get the percent full and inodes using `df`:

```
df  | grep '^/dev'
df  -i | grep '^/dev'
```

On my system that takes 4 milliseconds. The output would be easy to parse with a regex:

```
/dev/nvme0n1p1  20430864 11514536   8062444  59% /
/dev/nvme1n1    43169104 21146576  19804136  52% /data
```

**Comment by fopina** [comment]
Latest reply brought me back to this thread so I thought it'd be worth sharing for future people landing here: https://github.com/fopina/fluent-bit-in-diskfree

I believe I made a PR at some point to here to have the plugin merged upstream but there were some license issues plus different strategies.

In the meantime, feel free to use it as separate module (as it's faster than calling df and parsing as I initially posted here)
